File: android/src/main/java/com/compdfkitpdf/reactnative/util/CPDFSearchUtil.java
/*
 * Copyright © 2014-2025 PDF Technologies, Inc. All Rights Reserved.
 *
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE ComPDFKit LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 *
 */

package com.compdfkitpdf.reactnative.util;

import android.content.Context;
import android.util.Log;
import androidx.annotation.Nullable;
import com.compdfkit.core.document.CPDFDocument;
import com.compdfkit.core.page.CPDFPage;
import com.compdfkit.core.page.CPDFTextPage;
import com.compdfkit.core.page.CPDFTextRange;
import com.compdfkit.tools.common.views.pdfview.CPDFViewCtrl;
import com.compdfkit.ui.textsearch.CPDFTextSearcher;
import com.compdfkit.ui.textsearch.ITextSearcher;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import java.util.List;


public class CPDFSearchUtil {


  /**
   * Searches for keywords in the given PDF document using the specified text searcher.
   *
   * @param document       The PDF document to search within.
   * @param iTextSearcher  The text searcher to use for searching.
   * @param keywords       The keywords to search for.
   * @param searchOptions  Options for the search, such as case sensitivity and whole word matching.
   * @see com.compdfkit.core.page.CPDFTextSearcher.PDFSearchOptions#PDFSearchCaseInsensitive
   * @see com.compdfkit.core.page.CPDFTextSearcher.PDFSearchOptions#PDFSearchCaseSensitive
   * @see com.compdfkit.core.page.CPDFTextSearcher.PDFSearchOptions#PDFSearchMatchWholeWord
   * @see com.compdfkit.core.page.CPDFTextSearcher.PDFSearchOptions#PDFSearchConsecutive
   */
  public static WritableArray search(CPDFDocument document, ITextSearcher iTextSearcher, String keywords, int searchOptions) {
    Log.i("ComPDFKit", "Android --- Searching for keywords: " + keywords);
    WritableArray searchResults = Arguments.createArray();
    iTextSearcher.setSearchConfig(keywords, searchOptions);
    for (int pageIndex = 0; pageIndex < document.getPageCount(); pageIndex++) {
      CPDFPage page = document.pageAtIndex(pageIndex);
      List<CPDFTextRange> textRanges = iTextSearcher.searchKeyword(pageIndex);
      for (int i = 0; i < textRanges.size(); i++) {
        CPDFTextRange textRange = textRanges.get(i);
        WritableMap result = Arguments.createMap();
        result.putInt("pageIndex", page.getPageNum());
        result.putInt("location", textRange.location);
        result.putInt("length", textRange.length);
        result.putInt("textRangeIndex", i);
        searchResults.pushMap(result);
      }
    }
    Log.i("ComPDFKit", "Android --- Search completed with " + searchResults.size() + " results.");
    return searchResults;
  }

  public static void selection(Context context,  @Nullable CPDFViewCtrl pdfView, CPDFDocument document, int pageIndex, int textRangeIndex){
    ITextSearcher iTextSearcher = getTextSearcher(context, pdfView, document);
    iTextSearcher.searchBegin(pageIndex, textRangeIndex);
    if (pdfView != null) {
      pdfView.getCPdfReaderView().invalidateAllChildren();
    }
  }

  public static void clearSearch(Context context, @Nullable CPDFViewCtrl pdfView,  CPDFDocument document){
    ITextSearcher iTextSearcher = getTextSearcher(context, pdfView, document);
    iTextSearcher.cancelSearch();
    if (pdfView != null) {
      pdfView.getCPdfReaderView().invalidateAllChildren();
    }
  }

  public static String getText(CPDFDocument document, int pageIndex, int location, int length) {
    CPDFPage page = document.pageAtIndex(pageIndex);
    CPDFTextPage textPage = page.getTextPage();
    CPDFTextRange range = new CPDFTextRange(location, length);
    return textPage.getText(range);
  }

  private static ITextSearcher getTextSearcher(Context context, @Nullable CPDFViewCtrl pdfView, CPDFDocument document) {
    if (pdfView != null) {
      return pdfView.getCPdfReaderView().getTextSearcher();
    } else {
      return new CPDFTextSearcher(context, document);
    }
  }

}

File: ios/CPDFSearchUtil.swift
//
//  CPDFSearchUtil.swift
//  Pods
//
//  Created by kdanmobile on 2025/7/23.
//

import Foundation
import ComPDFKit
import ComPDFKit_Tools


class CPDFSearchUtil {
  
  static func searchText(from document : CPDFDocument?, keywords : String , options : CPDFSearchOptions) -> [[String: Any]] {
    var searchResults : [Dictionary<String, Any>] = []
    if(document == nil){
      return searchResults;
    }
    if let resultArray = document?.find(keywords, with: options) as? [[CPDFSelection]] {
      
      for array in resultArray {
        for selection in array {
          var dict: [String : Any] = [:]
          dict["pageIndex"] = selection.page.pageIndexInteger
          dict["location"] = selection.range.location
          dict["length"] = selection.range.length
          dict["textRangeIndex"] = array.firstIndex(of: selection)
          searchResults.append(dict)
          
        }
      }
    }
    return searchResults;
  }
  
  
  static func selection(from document : CPDFDocument?, info: NSDictionary) -> CPDFSelection? {
    guard
      let pageIndex = info["pageIndex"] as? Int,
      let location = info["location"] as? Int,
      let length = info["length"] as? Int
    else {
      return nil
    }
    
    guard let page = document?.page(at: UInt(pageIndex)) else {
      return nil
    }
    
    let range = NSRange(location: location, length: length)
    return page.selection(for: range)
  }
  
  static func getSearchText(from document: CPDFDocument?, pageIndex: Int, location: Int, length: Int) -> String? {
    guard let page = document?.page(at: UInt(pageIndex)) else {
      return nil
    }
    
    let range = NSRange(location: location, length: length)
    let selection = page.selection(for: range)
    return selection?.string();
  }
}
